### PR TITLE
ENH: Add support for Tuttifrutti diagnostic stack

### PR DESCRIPTION
**Pull request by slactjohnson**
## Description

Adds a module for generating TuttiFrutti classes and instantiating them.

Classes are dynamically created based on the specified diagnostics/devices. In theory there are about factorial(10) combinations of diagnostics/devices, so this seems like the way to go.

A factory function is provided for creating custom TuttiFrutti classes. Another function instantiates the device for you and returns it.

## Motivation and Context

The new laser control system uses TuttiFrutti diagnostic packages all over the place. Each TuttiFrutti is a special snowflake of diagnostics, tailored to the application. The suite of diagnostics used is standardized, and each TuttiFrutti is a single assembly, so it still makes sense to group them. 

## How Has This Been Tested?

Tested using devices in the B40 Heinz lab. Used as components in a forthcoming PR with MODS and TILE devices. 

Generates nice Typhos screens as well. 

## Where Has This Been Documented?

I wrote a long and somewhat silly docstring.

### CODE REVIEW

File: pcdsdevices/lasers/tuttifrutti.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        can use FCpt here to temporarily get around possible PV naming
+        incompatibilities with the L2SI laser PV naming convention (I'm looking
+        at you, EK9000...).
+
+    Examples
+    --------
+    # Create an object for a basic tuttifrutti containing NF/FF cameras,
+    # a spectrometer, and a filter slider.
+
+    ttf = TutttFruttiBase('LAS:TTF:01', nf=True, ff=True, spec=True, ell=True)
```
Comment (ZLLentz) [inline pcdsdevices/lasers/tuttifrutti.py:125]: I assume this is an example from an earlier version of the code
```suggestion
    ttf = TuttiFrutti('LAS:TTF:01', nf=True, ff=True, spec=True, ell=True)
```

File: pcdsdevices/lasers/tuttifrutti.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    from ophyd import Component as Cpt
+    from ophyd import FormattedComponent as FCpt
+
+    from fruit import Apple, Orange
+    apple = Cpt(Apple, ':APPLE', kind='normal') # Use TuttiFrutti base PV
+    orange = FCpt(Orange, 'IOC:LAS:ORANGE', kind='normal') # Use full base PV
+
+    dmisc = {'apple': apple, 'banana' : banana}
+
+    ttf = TutttFruttiBase('LAS:TTF:01', nf=True, ff=True, spec=True, ell=True,
```
Comment (ZLLentz) [inline pcdsdevices/lasers/tuttifrutti.py:138]:
```suggestion
    ttf = TuttiFrutti('LAS:TTF:01', nf=True, ff=True, spec=True, ell=True,
```

File: pcdsdevices/lasers/tuttifrutti.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    if pd:
+        raise(NotImplemented, "Pulse duration is not yet implemented")
+    if wfs:
+        raise(NotImplemented, "Wavefront sensor is not yet implemented")
+    if ell:
+        cpt = Cpt(Ell6, '_SL1:ELL:M1', kind='normal')
+        cpts['slider'] = cpt
+    if misc:  # This feels kind of hacky, but also kind of cool.
+        for cptname, cpt in misc.items():
+            cpts[cptname] = cpt
```
Comment (ZLLentz) [inline pcdsdevices/lasers/tuttifrutti.py:49]: Do you expect this to get a lot of use? Just curious. It seems fine, but I thought the Tuttifrutti were pretty modular so having a misc surprises me.

Comment (slactjohnson) [reply pcdsdevices/lasers/tuttifrutti.py:49]: I hope not. It just seemed like a nice convenience for dealing with off-normal deployments. If we start instantiating TuttiFruttis outside of the MODS, like they plan to in XPP, I can't guarantee that the components will be the same.

Comment (ZLLentz) [reply pcdsdevices/lasers/tuttifrutti.py:49]: Fair enough!

File: pcdsdevices/lasers/tuttifrutti.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        can use FCpt here to temporarily get around possible PV naming
+        incompatibilities with the L2SI laser PV naming convention (I'm looking
+        at you, EK9000...).
+
+    Examples
+    --------
+    # Create an object for a basic tuttifrutti containing NF/FF cameras,
+    # a spectrometer, and a filter slider.
+
+    ttf = TutttFrutti('LAS:TTF:01', nf=True, ff=True, spec=True, ell=True)
```
Comment (ZLLentz) [inline pcdsdevices/lasers/tuttifrutti.py:125]: Still have a typo here and below: triple t Tu`ttt`Frutti

File: pcdsdevices/lasers/tuttifrutti.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        cpt = Cpt(PCDSAreaDetectorTyphos, '_FF1:', kind='normal')
+        cpts['ff_camera'] = cpt
+    if spec:
+        cpt = Cpt(QminiSpectrometer, '_SP1', kind='normal')
+        cpts['spectrometer'] = cpt
+    if pm:
+        cpt = Cpt(El3174AiCh, '_PM1', kind='normal')
+        cpts['power_meter'] = cpt
+    if diode:
+        raise(NotImplemented, "Diode is not yet implemented")
```
Comment (klauer) [inline pcdsdevices/lasers/tuttifrutti.py:35]:
```suggestion
        raise NotImplementedError("Diode is not yet implemented")
```

File: pcdsdevices/lasers/tuttifrutti.py (hunk 4 above)
Comment (klauer) [inline pcdsdevices/lasers/tuttifrutti.py:125]:
```suggestion
    ttf = TuttiFrutti('LAS:TTF:01', nf=True, ff=True, spec=True, ell=True)
```

File: pcdsdevices/lasers/tuttifrutti.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    from ophyd import Component as Cpt
+    from ophyd import FormattedComponent as FCpt
+
+    from fruit import Apple, Orange
+    apple = Cpt(Apple, ':APPLE', kind='normal') # Use TuttiFrutti base PV
+    orange = FCpt(Orange, 'IOC:LAS:ORANGE', kind='normal') # Use full base PV
+
+    dmisc = {'apple': apple, 'orange' : orange}
+
+    ttf = TutttFrutti('LAS:TTF:01', nf=True, ff=True, spec=True, ell=True,
```
Comment (klauer) [inline pcdsdevices/lasers/tuttifrutti.py:138]:
```suggestion
    ttf = TuttiFrutti('LAS:TTF:01', nf=True, ff=True, spec=True, ell=True,
```

**Pull request merged**